### PR TITLE
Refactor ExportDialog pages into separate components

### PR DESCRIPTION
**Pull request by cpirich**
* Recent reviews of the `ExportDialog` have suggested that we refactor the single `ExportDialog` class.
* This PR splits the dialog pages (Intro, Platform, Icon, Publish, Generating) into separate React components. Note that the `IconPage` is still disabled, but we do plan to bring it back in the future, so it is included in the refactor.

### CODE REVIEW

**Comment by jmkulwik** [comment]
Can you split this into separate commits?
- Move code in one commit
- Change code in the other commit
Otherwise the reviewers will need to manually put this in a diff tool to do a review

**Comment by cpirich** [comment]
> Can you split this into separate commits?
> 
> * Move code in one commit
> * Change code in the other commit
>   Otherwise the reviewers will need to manually put this in a diff tool to do a review

I wanted to make sure that the code compiles properly with each commit, which I believe is a good general practice. I already looked at the diff in github and don't think it is too hard to sort out.

**Comment by cpirich** [comment]
All pages except the `GeneratingPage` are straightforward copies of the `renderXYZPage()` method into a new class.

The `GeneratingPage` just moves to using `props` instead of `state` (and therefore, has `propTypes`)

The refactoring of styles so they can be shared (into `styles.js`) is a little hard to compare no matter how I sequence the commits.

File: apps/src/code-studio/components/ExportDialog/Dialog.jsx [hunk 1]
```diff
@@ -73,122 +51,35 @@ const styles = {
...
-  iosAppStoreButton: {
-    ...baseStyles.button,
-    backgroundColor: color.purple,
-    color: color.white
-  },
-  androidGooglePlayButton: {
-    ...baseStyles.button,
-    backgroundColor: color.purple,
-    color: color.white
-  },
```
Comment (jmkulwik) [inline apps/src/code-studio/components/ExportDialog/Dialog.jsx:90]: Thanks for cleaning these two up.

File: apps/src/code-studio/components/ExportDialog/Dialog.jsx [hunk 2]
```diff
@@ -73,122 +51,35 @@ const styles = {
...
   },
   backButtonDisabled: {
-    ...baseStyles.button,
+    ...commonStyles.button,
     backgroundColor: color.gray,
     color: color.white
   },
-  p: {
-    ...baseStyles.text
-  },
```
Comment (jmkulwik) [inline apps/src/code-studio/components/ExportDialog/Dialog.jsx:118]: nice cleanup 
(some of these comments are for my own notes. I'll try to make it obvious if any action is needed)

**Comment by jmkulwik** [comment]
Looks great! Thanks for doing this cleanup! A couple more notes:
1. Add some documentation - What's an apk? What is Expo? Why do we need to use these patterns? Add a note explaining why we expect a 15 minute wait time.
2. In the render method: The {blockExport && } section followed by the {!blockExport && } section worries me a bit. I could see someone in the future rewriting parts of these conditionals without realizing this is, in essence, an if/else. Factor this out so it is actually an if/else so it will be clearer to the next editor.
3. This can be refactored further. i.e. since ExportDialog is now a container for a series of dialogs, it no longer makes sense to have the apk and expo actions as part of ExportDialog. They should be part of a child component (the generating component, perhaps? or maybe a new component?). ExportDialog is still doing too much.

**Comment by cpirich** [comment]
> Looks great! Thanks for doing this cleanup! A couple more notes:
> 
> 1. Add some documentation - What's an apk? What is Expo? Why do we need to use these patterns? Add a note explaining why we expect a 15 minute wait time.

Can do.
> 2. In the render method: The {blockExport && } section followed by the {!blockExport && } section worries me a bit. I could see someone in the future rewriting parts of these conditionals without realizing this is, in essence, an if/else. Factor this out so it is actually an if/else so it will be clearer to the next editor.

Yes, agreed. This was my attempt to simulate the crazy logic in `ShareAllowedDialog` (see references to `this.sharingDisabled()`)
> 3. This can be refactored further. i.e. since ExportDialog is now a container for a series of dialogs, it no longer makes sense to have the apk and expo actions as part of ExportDialog. They should be part of a child component (the generating component, perhaps? or maybe a new component?). ExportDialog is still doing too much.

This is a much longer discussion. The state machine that drives the export exists outside of the pages. It is quite complex to tease apart. The page components were relatively easy to refactor since they are primarily used for rendering only. To go further, we'd probably want to split `Dialog.jsx` into its own rendering component and some sort of `export state machine` that is independent. I certainly think the work in this PR is a step in the right direction. I'll need to think further on a bigger re-architecture.

**Comment by jmkulwik** [comment]
> This is a much longer discussion. The state machine that drives the export exists outside of the pages. It is quite complex to tease apart. The page components were relatively easy to refactor since they are primarily used for rendering only. To go further, we'd probably want to split `Dialog.jsx` into its own rendering component and some sort of `export state machine` that is independent. I certainly think the work in this PR is a step in the right direction. I'll need to think further on a bigger re-architecture.

Agreed. I'm fine skipping this one for now.

**Comment by cpirich** [comment]
@jmkulwik Some docs added in a comment block and the `blockExport` path is now factored out per your request. Please take a look.

File: apps/src/code-studio/components/ExportDialog/Dialog.jsx [hunk 3]
```diff
@@ -562,9 +619,10 @@ class ExportDialog extends React.Component {
     const needToSignIn =
       !isProjectLevel && signInState !== SignInState.SignedIn;
     const blockExport = userSharingDisabled || needToSignIn;
-    const blockText = userSharingDisabled
-      ? i18n.sharingBlockedByTeacher()
-      : i18n.createAccountToShareDescription();
+
+    if (blockExport) {
+      return this.renderWrappedInBaseDialog(this.renderBlocked());
```
Comment (jmkulwik) [inline apps/src/code-studio/components/ExportDialog/Dialog.jsx:624]: NIT: `this.renderBlocked()` reads like a boolean. Perhaps rename to something like `renderBlockedMessage()`

**Review by jmkulwik — APPROVED** [review]
Thank you for doing this refactor!

**Pull request merged**